core/domain/feature_gating_services.py

```diff
@@ -0,0 +1,191 @@
+# coding: utf-8
+#
+# Copyright 2020 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""The service for gating features.
+
+This service provides two interfaces to access the feature flag values for
+clients and servers respectively.
+
+For clients, please use 'evaluate_all_feature_flag_value_for_client' from
+request handlers with client context.
+
+For servers, please directly call 'evaluate_feature_flag_value_for_server' with
+the name of the feature.
+
+For more details of the usage of these two method, please refer their docstrings
+in this file.
+"""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+from constants import constants
+from core import platform_feature_list
+from core.domain import platform_parameter_domain as param_domain
+from core.domain import platform_parameter_registry as registry
+
+
+ALL_FEATURES_LIST = (
+    platform_feature_list.DEV_FEATURES_LIST +
+    platform_feature_list.TEST_FEATURES_LIST +
+    platform_feature_list.PROD_FEATURES_LIST
+)
+
+ALL_FEATURES_NAMES_SET = set(ALL_FEATURES_LIST)
+
+
+def create_evaluation_context_for_client(client_context_dict):
+    """Returns context instance for evaluation, using the information
+    provided by clients.
+
+    Args:
+        client_context_dict: dict. The client side context.
+
+    Returns:
+        EvaluationContext. The context for evaluation.
+    """
+    return param_domain.EvaluationContext.from_dict(
+        client_context_dict,
+        {
+            'server_mode': _get_server_mode()
```

Comment:
is this key `server_mode` used in other files? (I'm assuming it to be defined in param_domain if so should we use the constant instead of raw_string?)

EDIT: I see this string is used in other places, so maybe define this in a constant?